src/app/template/template-sidebar/containers/tags.container.ts

```diff
@@ -23,7 +26,47 @@ export class TagsContainerComponent {
   constructor(private store: Store<State>) {
   }
 
-  public update() {
-    this.store.dispatch(new templateActions.LoadTemplatesRequest())
+  public editTag(tagEdit: TagEditAction) {
+    this.template$.take(1).subscribe((template: Template) => {
+      const newTag: Tag = {
+        resourceid: template.id,
+        resourcetype: 'Template',
+        key: tagEdit.newTag.key,
+        value: tagEdit.newTag.value,
+        account: template.account,
+        domain: template.domain,
+        domainid: template.domainId
+      };
+      const newTags: Tag[] = template.tags.filter(t => tagEdit.oldTag.key !== t.key);
+      newTags.push(newTag);
+      const result = { ...template, tags: newTags };
+      this.store.dispatch(new templateActions.UpdateTemplate(result));
+    });
+  }
+
+  public deleteTag(tag: Tag) {
+    this.template$.take(1).subscribe((template: Template) => {
+      const newTags = template.tags.filter(_ => tag.key !== _.key);
+      this.store.dispatch(new templateActions.UpdateTemplate({ ...template, tags: newTags }));
+    });
+  }
+
+  public addTag(keyValuePair: KeyValuePair) {
+    this.template$.take(1).subscribe((template: Template) => {
+      const newTag = {
+        resourceid: template.id,
+        resourcetype: 'SecurityGroup',
```

Comment:
Why above you use ` resourcetype: 'Template'` but there is a `resourcetype: 'SecurityGroup'`?